src/vm/memory.zig

```diff
@@ -104,71 +142,232 @@ pub const Region = struct {
+    // Executable sbpf_version
+    version: sbpf.Version,
+    // VM configuration
+    config: *const exe.Config,
+
+    // CoW callback
+    //cow_cb: ?MemoryCowCallback,
+    // Cache of the last `MappingCache::SIZE` vm_addr => region_index lookups
+    //cache: MappingCache,
+
+    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L241
+    fn init(
+        allocator: std.mem.Allocator,
+        _regions: []const Region,
+        version: sbpf.Version,
+        config: *const exe.Config,
+    ) (error{OutOfMemory} || InitError)!UnalignedMemoryMap {
+        const sorted_regions = try allocator.dupe(Region, _regions);
+        defer allocator.free(sorted_regions);
+        std.mem.sort(Region, sorted_regions, {}, Region.lessThanFn);
+
+        if (Region.regionsOverlap(sorted_regions)) return error.InvalidMemoryRegion;
+
+        const region_addresses = try allocator.alloc(u64, sorted_regions.len);
+        errdefer allocator.free(region_addresses);
+
+        const regions = try allocator.alloc(Region, _regions.len);
+        errdefer allocator.free(regions);
+
+        var self: UnalignedMemoryMap = .{
+            .regions = regions,
+            .region_addresses = region_addresses,
+            // .cache
+            .version = version,
+            .config = config,
+            // .cow_cb
+        };
+
+        _ = self.constructEytzingerOrder(sorted_regions, 0, 0);
+
+        return self;
+    }
+
+    fn deinit(self: *const UnalignedMemoryMap, allocator: std.mem.Allocator) void {
+        allocator.free(self.regions);
+        allocator.free(self.region_addresses);
+    }
+
+    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L218
+    fn constructEytzingerOrder(
+        self: UnalignedMemoryMap,
+        ascending_regions: []Region,
+        _in_index: usize,
+        out_index: usize,
+    ) usize {
+        if (out_index >= ascending_regions.len) return _in_index;
+
+        const in_index = self.constructEytzingerOrder(
+            ascending_regions,
+            _in_index,
+            (out_index *| 2) +| 1,
+        );
+
+        {
+            self.regions[out_index] = ascending_regions[in_index];
+            // agave mutates ascending_regions like this, but there's no reason for it (mem::take).
+            // ascending_regions[in_index] = Region.DEFAULT; // agave
+        }
+
+        self.region_addresses[out_index] = self.regions[out_index].vm_addr_start;
+
+        return self.constructEytzingerOrder(
+            ascending_regions,
+            in_index +| 1,
+            (out_index *| 2) +| 2,
+        );
+    }
+
+    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L293
+    fn find_region(self: *const UnalignedMemoryMap, vm_addr: u64) ?Region {
+        // NOTE: agave-like cache unimplemented. Does not seem necessary.
```

Comment:
```suggestion
    // [agave] https://github.com/anza-xyz/sbpf/blob/a8247dd30714ef286d26179771724b91b199151b/src/memory_region.rs#L293
    // NOTE: agave-like cache unimplemented. Does not seem necessary.
    fn find_region(self: *const UnalignedMemoryMap, vm_addr: u64) ?Region {
```

Reply:
https://github.com/Syndica/sig/pull/639/commits/e2a0e3c58e0f908854d59b9212eca28bc5f1bd0e#diff-23e4e8b8c3b0ebce6d44fff7c7ff93546043dcee6a94c6c2b54d5482d0457468R334-R339 moved the comment here